accepted/2021/default_string_globalization.md

```diff
@@ -0,0 +1,390 @@
+# Improving the _string_ developer experience
+
+Questions? Contact: @tarekgh @safern @GrabYourPitchforks
+
+## Problem summary
+
+APIs on `string` are not consistently ordinal or linguistic, and this leads to confusion among our developer community and bugs in applications. https://github.com/dotnet/runtime/issues/43956 describes this pain in more detail, and it aggregates issues where developers experienced breaks when upgrading from .NET Core 3.1 to .NET 5 due to runtime changes exposing latent bugs in their own applications.
+
+That GitHub issue describes various options we were considering, including relying solely on Roslyn analyzers or providing runtime compatibility switches. None of these issues is very palatable because they either force the developers to think about globalization concerns or they make it difficult for library authors to reason about what behaviors their components will exhibit at runtime.
+
+This draft proposes an alternate solution which attempts to provide the best of all worlds: steering new developers toward success, letting people with existing code bases know when they have issues, and allowing existing compiled DLLs to work largely as expected.
+
+## Characteristics of an ideal solution
+
+An ideal solution to this problem should have two qualities.
+
+1. `string` and related APIs should default to ordinal behavior unless the caller makes an explicit request for a different behavior.
+
+2. Developers should not be exposed to any globalization-related concepts unless they explicitly seek globalization support.
+
+The first characteristic is fairly straightforward. APIs like `string.Compare`, `string.Contains`, `string.Equals`, `string.IndexOf`, and others should all behave in an ordinal fashion. Developers shouldn't have to rely on Roslyn analyzers or MSDN to know what behavior a given API has.
+
+The second characteristic requires a bit more explanation. Popular programming languages like Java, JavaScript, Go, Python, C/C++, and others treat string APIs as ordinal. .NET is unique in that we build globalization concepts directly into many basic `string` APIs. Unfortunately, evidence shows that .NET's behavior in this regard is largely unexpected and confusing to our developer audience.
+
+[Java 16's `String` class](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/String.html) (which is the closest to .NET's `string` class) is a good example of this principle. Per their own documentation:
```

Comment:
It's good that we're looking to learn from other ecosystems. Is there anything we can learn from the other ecosystems you mention above, beyond your note that they separate string from globalization? (Go, Rust etc)

Reply:
Oh, absolutely. Most other ecosystems also enforce opinionated behaviors regarding UTF-8 string well-formedness. But that continues to be a point of contention for us. ;)

Reply:
I think an important thing to keep in mind about scenarios like `Rust` is that they design most of their types around the principle of "zero cost abstraction": <https://doc.rust-lang.org/book/ch00-00-introduction.html#people-who-value-speed-and-stability>
> Rust is for people who crave speed and stability in a language. By speed, we mean the speed of the programs that you can create with Rust and the speed at which Rust lets you write them. The Rust compiler’s checks ensure stability through feature additions and refactoring. This is in contrast to the brittle legacy code in languages without these checks, which developers are often afraid to modify. By striving for zero-cost abstractions, higher-level features that compile to lower-level code as fast as code written manually, Rust endeavors to make safe code be fast code as well.

So while things like `str` are always sliced to valid UTF-8 (unless using one of the `unchecked` slicing methods), they ensure the relevant other components are exposed to allow working with and dealing with "raw parts", that is the `bytes` rather than `runes` (to put it into .NET terms) so power users have the ability to do things themselves when and as appropriate.

Likewise, I don't believe rust has a case insensitive comparer in `std` today. You need to pull it in through an external crate (basically `nupkg`) and specify it as the comparer to use in your `match` expressions (in this way, its more like `IComparer`, but where the implementation is a `struct` and done via generic specialization so its "zero cost" and inlining/etc happen).